Create calculatePoseFromTarget

alicekuznetsov opened the pull request:
Reasons why this _might_ not work:
- result.getTimestampSeconds(); is not synced with Timer.getFPGATimestamp();
- I'm not smart
- Also, chasing the target won't work if we're constantly resetting our position based on the target we're chasing, so remember that ig

alicekuznetsov committed:
I did a thing??

Reasons why this *might* not work:
- result.getTimestampSeconds(); is not synced with Timer.getFPGATimestamp();
- I'm not smart
- Also, chasing the target won't work if we're constantly resetting our position based on the target we're chasing, so remember that ig

IanSzalai commented:
I'm not really sure how `result.getTimestampSeconds()` differs from `Timer.getFPGATimestamp()` but worse case we can purely use the former and it should be fine.

IanSzalai commented:
Ideally, the vision subsystem has the bulk of the vision processing code, rather than a command. This would distill into some `getRobotPose()` (or similarly named) method that is used in the `UpdatePoseEstimator` command. To my understanding, what complicates this is that if there is no valid pose, you don't want to add a vision measurement. I see two ways around this.

1. Simple include another getter method in the vision system that will simply return true if the vision data is good to use.
```java
if (subVision.isPoseValid()) {
  addVisionMeasurement(subVision.getRobotPose());
}
// method names for example
```
2. Return null from `subVision.getRobotPose()` if the data isn't good to use. Then you could either check for null in `UpdatePoseEstimator` or in `subDrivetrain.addVisionMeasurement()` eg
```java
public void addVisionMeasurement(Pose2d visionRobotPose, double timestampSeconds) {
  if (visionRobotPose != null) {
    poseEstimator.addVisionMeasurement(
        visionRobotPose,
        timestampSeconds);
  }
}
```
or
```java
// execute method of UpdatePoseEstimator
public void execute() {
  subDrivetrain.updatePoseEstimator();
  if (subVision.getRobotPose() != null) {
    subDrivetrain.addVisionMeasurement(subVision.getRobotPose(), subVision.getRobotPoseTimestamp());
  }
}
```

I may be misunderstanding the problem at hand, please let me know if that's the case.

IanSzalai commented on src/main/java/frc/robot/commands/getInRangeOfTag.java:
```diff
@@ -23,6 +23,8 @@
 import frc.robot.subsystems.Vision;
 
 // Note: This command assumes that you know where you are on the field, since every other position is relative to you
+// Ok so I kinda wanna throw this command into a fire and start over so I understand it better B)
+// Also it would just be better and I kinda need to anyway
```
i agree

IanSzalai commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -285,10 +285,10 @@ public void updatePoseEstimator() {
    */
-  public void addVisionMeasurement(Pose2d visionRobotPose) {
+  public void addVisionMeasurement(Pose2d visionRobotPose, double timestampSeconds) {
     poseEstimator.addVisionMeasurement(
         visionRobotPose,
-        Timer.getFPGATimestamp());
+        timestampSeconds);
```
assuming you're only calling this method with valid pose data, does this change only matter for accuracy beyond +/- 20ms (50hz)?

alicekuznetsov replied:
I just thought that a timestamp associated with the vision data measurement would be more accurate than a timestamp that's run slightly after the vision data. I don't think the increase in accuracy is too influential, but I think it's still better to do it like this.

IanSzalai replied:
i agree. the theoretical worst delta is 20ms (if i understand this correctly) and that could cause noticeable error

alicekuznetsov committed:
Removed stuff and moved logic to Vision

also please read my comment on line 47 of vision I literally just need a variable type because dictionary is obsolete

IanSzalai commented:
new changes look good to me. do you think that this structure makes the most sense?

alicekuznetsov committed:
Added poseEstimationType

alicekuznetsov commented:
> new changes look good to me. do you think that this structure makes the most sense?

I think so. Combining all of the types of pose estimators into one command but keeping them separate in subsystems seems smart. 
Not having real positions for each apriltag (ie. every tag is at 0 rn) is going to make the whole process a bit weird, but that sounds like a problem for future me 

Just committed changes that should (?) fix the "chasing tag" problem. Correct me if we shouldn't use it or if I did it wrong. If we decide to implement a chasing tag code, I will also add a filter on the specific tag that we use to make its pose invalid.

alicekuznetsov committed:
Removed unused imports

alicekuznetsov committed:
Added multiple target pose estimation

Pretty confident that the position stuff works. Not so confident about the rotation. But you gotta yolo it sometimes

This is basically the easiest way of doing it; adding together your estimated poses and getting the average from that

IanSzalai committed:
Merge branch 'main' of https://github.com/IanSzalai/SuperSwerve into calculatePoseFromVision

IanSzalai committed:
h

alicekuznetsov committed:
ok so apparently i do need to use quaternions

IanSzalai committed:
Merge branch 'main' into calculatePoseFromVision

IanSzalai committed:
fix merge

alicekuznetsov committed:
awfiwagiojawgawfqwfa

IanSzalai committed:
wip

IanSzalai committed:
make code more readable

IanSzalai committed:
put tag in middle of field and put pose2d on dash

IanSzalai committed:
removed unused imports

IanSzalai committed:
removed pose estimation type enum

IanSzalai committed:
Merge branch 'main' of https://github.com/IanSzalai/SuperSwerve into calculatePoseFromVision

IanSzalai committed:
Merge branch 'main' of https://github.com/IanSzalai/SuperSwerve into remove-enum

alicekuznetsov committed:
Merge pull request #21 from IanSzalai/remove-enum

removed pose estimation type enum

alicekuznetsov committed:
Made useVision false, will make true in 2023
